src/preview/rewriteStyleSheet.test.ts

```diff
@@ -243,41 +243,74 @@ describe("rewriteStyleSheet", () => {
+    expect(sheet.cssRules[0].selectorText).toEqual(":not(:hover), :not(.pseudo-hover), :not(.pseudo-hover-all *)")
   })
 
-  it('supports ":not"', () => {
+  it('supports ":not" in shadow DOM', () => {
     const sheet = new Sheet(":not(:hover) { color: red }")
+    rewriteStyleSheet(sheet as any, true)
+    expect(sheet.cssRules[0].selectorText).toEqual(":not(:hover), :not(.pseudo-hover), :not(:host(.pseudo-hover-all) *)")
+  })
+
+  it('supports complex use of ":not"', () => {
+    const sheet = new Sheet("foo:focus:not(:hover, .bar:active) .baz { color: red }")
     rewriteStyleSheet(sheet as any)
-    expect(sheet.cssRules[0].selectorText).toEqual(":not(:hover), :not(.pseudo-hover)")
+    const selectors = sheet.cssRules[0].getSelectors()
+    expect(selectors).toContain("foo:focus:not(:hover, .bar:active) .baz")
+    expect(selectors).toContain("foo.pseudo-focus:not(.pseudo-hover, .bar.pseudo-active) .baz")
+    expect(selectors).toContain(".pseudo-focus-all foo:not(.pseudo-hover-all *, .pseudo-active-all .bar) .baz")
+  })
+
+  it('supports complex use of ":not" in shadow DOM', () => {
+    const sheet = new Sheet("foo:focus:not(:hover, .bar:active) .baz { color: red }")
+    rewriteStyleSheet(sheet as any, true)
+    const selectors = sheet.cssRules[0].getSelectors()
+    expect(selectors).toContain("foo:focus:not(:hover, .bar:active) .baz")
+    expect(selectors).toContain("foo.pseudo-focus:not(.pseudo-hover, .bar.pseudo-active) .baz")
+    expect(selectors).toContain(":host(.pseudo-focus-all) foo:not(:host(.pseudo-hover-all) *, :host(.pseudo-active-all) .bar) .baz")
+  })
+
+  it('supports ":not" inside ":host"', () => {
+    const sheet = new Sheet(":host(.foo:not(:hover)) .baz:active { color: red }")
+    rewriteStyleSheet(sheet as any, true)
+    const selectors = sheet.cssRules[0].getSelectors()
+    expect(selectors).toContain(":host(.foo:not(:hover)) .baz:active")
+    expect(selectors).toContain(":host(.foo:not(.pseudo-hover)) .baz.pseudo-active")
+    expect(selectors).toContain(":host(.foo:not(.pseudo-hover-all).pseudo-active-all) .baz")
+  })
+
+  it('supports ":not" inside and outside of ":host"', () => {
+    const sheet = new Sheet(":host(.foo:not(:hover)) .baz:not(:active) { color: red }")
+    rewriteStyleSheet(sheet as any, true)
+    const selectors = sheet.cssRules[0].getSelectors()
+    expect(selectors).toContain(":host(.foo:not(:hover)) .baz:not(:active)")
+    expect(selectors).toContain(":host(.foo:not(.pseudo-hover)) .baz:not(.pseudo-active)")
+    expect(selectors).toContain(":host(.foo:not(.pseudo-hover-all)) .baz:not(:host(.pseudo-active-all) *)")
   })
 
   it('supports ":has"', () => {
     const sheet = new Sheet(":has(:hover) { color: red }")
     rewriteStyleSheet(sheet as any)
-    expect(sheet.cssRules[0].cssText).toEqual(":has(:hover), :has(.pseudo-hover) { color: red }")
+    expect(sheet.cssRules[0].cssText).toEqual(":has(:hover), :has(.pseudo-hover), .pseudo-hover-all :has(*) { color: red }")
```

Comment:
Now we also emit a working rule with `.pseudo-*-all` for input rules that have `:has(<pseudo-state>)`.